File: engine/julia/exotic_multileg_ladder_engine.jl
module ExoticMultiLegLadderEngineModule

export ExoticMultiLegLadderState, construct_strip, construct_strap

struct ExoticMultiLegLadderState
    strike_rung1::Float64
    strike_rung2::Float64
    strike_rung3::Float64
    lambda_elasticity::Float64
    net_package_premium::Float64
    strategy_archetype::Int32
end

function construct_strip(spot::Float64, atm::Float64, call_prem::Float64, put_prem::Float64)::ExoticMultiLegLadderState
    tot = (2.0 * put_prem) + call_prem
    delta = (1.0 * 0.50) + (2.0 * (-0.50))
    lam = tot > 0.001 ? (delta * spot) / tot : 0.0
    return ExoticMultiLegLadderState(atm, atm, atm, lam, tot, 1)
end

function construct_strap(spot::Float64, atm::Float64, call_prem::Float64, put_prem::Float64)::ExoticMultiLegLadderState
    tot = (2.0 * call_prem) + put_prem
    delta = (2.0 * 0.50) + (1.0 * (-0.50))
    lam = tot > 0.001 ? (delta * spot) / tot : 0.0
    return ExoticMultiLegLadderState(atm, atm, atm, lam, tot, 2)
end

end

File: engine/rust/src/exotic_multileg_ladder_engine.rs
#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct ExoticMultiLegLadderState {
    pub strike_rung1: f64,
    pub strike_rung2: f64,
    pub strike_rung3: f64,
    pub strike_rung4: f64,
    pub lambda_elasticity: f64,
    pub net_package_premium: f64,
    pub max_sweet_spot_profit: f64,
    pub strategy_archetype: u16,
    pub call_legs_count: u8,
    pub put_legs_count: u8,
    pub _padding: [u8; 4],
}

impl ExoticMultiLegLadderState {
    pub fn new_strip(spot: f64, atm: f64, call_prem: f64, put_prem: f64) -> Self {
        let total_prem = (2.0 * put_prem) + call_prem;
        let delta = (1.0 * 0.50) + (2.0 * (-0.50));
        let lam = if total_prem > 0.001 { (delta * spot) / total_prem } else { 0.0 };

        Self {
            strike_rung1: atm,
            strike_rung2: atm,
            strike_rung3: atm,
            strike_rung4: 0.0,
            lambda_elasticity: lam,
            net_package_premium: total_prem,
            max_sweet_spot_profit: 999999.0,
            strategy_archetype: 1,
            call_legs_count: 1,
            put_legs_count: 2,
            _padding: [0; 4],
        }
    }
}
